qualtran/bloqs/phase_estimation/lp_resource_state.py

```diff
@@ -0,0 +1,179 @@
+from qualtran.cirq_interop import CirqGateAsBloq
+from qualtran.cirq_interop.t_complexity_protocol import TComplexity
+
+if TYPE_CHECKING:
+    from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+
+@attrs.frozen
+class _LPRSHelper(GateWithRegisters):
+    """Helper Bloq to prepare an intermediate resource state which can be used in AA
+
+    Prepares the intermediate resource state described in Eq 19 of
+    https://arxiv.org/pdf/1805.03662.pdf, which can then be used in a single round of
+    Amplitude Amplification to boost the amplitude of desired resource state to 1.
+    """
+
+    bitsize: int
+    eps: float = 1e-11
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature.build(m=self.bitsize, anc=1)
+
+    def short_name(self) -> str:
+        return 'LPRS'
+
+    def decompose_from_registers(
+        self, *, context: cirq.DecompositionContext, **quregs: NDArray[cirq.Qid]
+    ) -> cirq.OP_TREE:
+        q, anc = quregs['m'].tolist()[::-1], quregs['anc']
+        yield [OnEach(self.bitsize, Hadamard()).on(*q), Hadamard().on(*anc)]
+        for i in range(self.bitsize):
+            rz_angle = -2 * np.pi * (2**i) / (2**self.bitsize + 1)
+            yield Rz(angle=rz_angle).controlled().on(q[i], *anc)
+        yield Rz(angle=-2 * np.pi / (2**self.bitsize + 1)).on(*anc)
+        yield Hadamard().on(*anc)
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        return {
+            (
+                CZPowGate(exponent=1 / ((2**self.bitsize + 1) * np.pi), global_shift=-0.5),
+                self.bitsize,
+            ),
+            (ZPowGate(exponent=1 / ((2**self.bitsize + 1) * np.pi), global_shift=-0.5), 1),
+            (Hadamard(), 2 + self.bitsize),
+        }
+
+    def _t_complexity_(self) -> 'TComplexity':
+        return TComplexity(rotations=self.bitsize + 1, clifford=2 + self.bitsize)
+
+
+@attrs.frozen
+class LPResourceState(GateWithRegisters):
+    r"""Prepares optimal resource state $\chi_{m}$ proposed by A. Luis and J. Peřina (1996)
+
+    Uses a single round of amplitude amplification, as described in Ref[2], to prepare the
+    resource state from Ref[1] described as
+
+    $$
+    \chi_{m} = \sqrt{\frac{2}{2^m + 1}}\sum_{n=0}^{2^m - 1}\sin{\frac{\pi(n+1)}{2^m+1}}|n\rangle
+    $$
+
+    Args:
+        bitsize: The size of the phase register to prepare the resource state on.
+
+
+    References:
+        1. [Optimum phase-shift estimation and the quantum description of the phase
+        difference](https://journals.aps.org/pra/abstract/10.1103/PhysRevA.54.4564)
+        2. [Encoding Electronic Spectra in Quantum Circuits with Linear T
+        Complexity](https://arxiv.org/abs/1805.03662) Section II-B
+    """
+
+    bitsize: int
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('m', QUInt(self.bitsize), side=Side.RIGHT)])
+
+    def state(self):
```

Comment:
Should this be apart of the class? Maybe it should be called expected_state. Also docstring + return type hint.

Reply:
It's only used for testing right?

Reply:
Yes, I can remove it and move to the test file

Reply:
done